### PR TITLE
Array parser refactoring

### PR DESCRIPTION
**Pull request by Tigrov**
Reduce executed code and checks

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | -

### CODE REVIEW

**Comment by Tigrov** [comment]
Also suggest new design for next major version
https://github.com/yiisoft/db-pgsql/blob/dc74029c9031102fe6d089e2ad615cb626ee5b81/src/ArrayParser.php
https://github.com/yiisoft/db-pgsql/blob/dc74029c9031102fe6d089e2ad615cb626ee5b81/src/ColumnSchema.php#L110

**Comment by terabytesoftw** [comment]
> Also suggest new design for next major version https://github.com/yiisoft/db-pgsql/blob/dc74029c9031102fe6d089e2ad615cb626ee5b81/src/ArrayParser.php
> 
> https://github.com/yiisoft/db-pgsql/blob/dc74029c9031102fe6d089e2ad615cb626ee5b81/src/ColumnSchema.php#L110

We can do this package is independent of the db version, so there would be no problem, if the improvement is good.

**Review by terabytesoftw — APPROVED** [review]

**Comment by Tigrov** [comment]
Or better make it callable:
`$value = (new ArrayParser($value))();`

File: tests/ArrayParserTest.php [hunk 1]
```diff
@@ -27,7 +27,7 @@ public function testParser(): void
         $this->assertSame([[0 => 'text'], [0 => null], [0 => '1']], $arrayParse->parse('{{text},{NULL},{1}}'));
         $this->assertSame(
             [[0 => '[",","null",true,"false","f"]'], 1 => ''],
-            $arrayParse->parse('"{"[\",\",\"null\",true,\"false\",\"f\"]"}"')
+            $arrayParse->parse('{{"[\",\",\"null\",true,\"false\",\"f\"]"},""}')
```
Comment (darkdef) [inline tests/ArrayParserTest.php:30]: You change working tests. Why?
If need - add new testcase, without remove old.

**Comment by darkdef** [comment]
> @samdark @darkdef What do you think, this proposal, we can implement it, but with all the accumulated improvements, releases a mayor version.

Better - redesign of ColumnTypes in 2.* (without phptypecast/dbtypecast with many "if")

File: tests/ArrayParserTest.php (hunk 1 above)
Comment (Tigrov) [reply tests/ArrayParserTest.php:30]: `'"{"[\",\",\"null\",true,\"false\",\"f\"]"}"'` it is incorrect value for the test.
`'{{"[\",\",\"null\",true,\"false\",\"f\"]"},""}'` it is correct.

Comment (Tigrov) [reply tests/ArrayParserTest.php:30]: Even `'{{"[\",\",\"null\",true,\"false\",\"f\"]"},""}'` is incorrect due to the value is multi-dimensional or one-dimensional. But for the test it is ok.

Comment (darkdef) [reply tests/ArrayParserTest.php:30]: How about have splitted test for one- and many-dimensional arrays?

Comment (Tigrov) [reply tests/ArrayParserTest.php]: Ok, good.

File: tests/ArrayParserTest.php [hunk 2]
```diff
@@ -25,9 +25,10 @@ public function testParser(): void
         $this->assertSame([0 => '1', 1 => '2', 2 => '3'], $arrayParse->parse('{1,2,3}'));
         $this->assertSame([0 => '1', 1 => '-2', 2 => null, 3 => '42'], $arrayParse->parse('{1,-2,NULL,42}'));
         $this->assertSame([[0 => 'text'], [0 => null], [0 => '1']], $arrayParse->parse('{{text},{NULL},{1}}'));
+        $this->assertSame([0 => ''], $arrayParse->parse('{""}'));
         $this->assertSame(
-            [[0 => '[",","null",true,"false","f"]'], 1 => ''],
-            $arrayParse->parse('"{"[\",\",\"null\",true,\"false\",\"f\"]"}"')
+            [[0 => '[",","null",true,"false","f"]']],
```
Comment (darkdef) [inline tests/ArrayParserTest.php:30]: add parse test for complex string: $arrayParse->parse('"{"[\",\",\"null\",true,\"false\",\"f\"]"}"')

Comment (Tigrov) [reply tests/ArrayParserTest.php:30]: This is incorrect test. PostgreSQL will never return string `'"{"[",","null",true,"false","f"]"}"'` for `array` value. It shold be starts with `{` and ends with `}`. In the example it starts and ends with `"`.

Comment (darkdef) [reply tests/ArrayParserTest.php:30]: If i save to column 'varchararray_col' this data: new ArrayExpression([0 => '[",","null",true,"false","f"]'])
then after read i have this: string(41) "{"[\",\",\"null\",true,\"false\",\"f\"]"}"
Please add test for parse this string

Comment (darkdef) [reply tests/ArrayParserTest.php:30]: https://github.com/yiisoft/db-pgsql/tree/arr_test

Comment (Tigrov) [reply tests/ArrayParserTest.php]: Ok, added as additional array value.

**Review by darkdef — APPROVED** [review]
Thanks.

**Comment by darkdef** [comment]
Please add line to changelog

**Comment by Tigrov** [comment]
Added lines to changelog

**Comment by darkdef** [comment]
Thanks!
